src/netconfig.cpp

```diff
@@ -0,0 +1,400 @@
+    const char* property;
+    if (isStatic)
+    {
+        ++args;
+        property = Dbus::ethStNameServers;
+    }
+    else
+    {
+        property = Dbus::ethNameServers;
+    }
+
+    const auto [_, srv] = args.asIpAddress();
+    args.expectEnd();
+
+    printf("%s DNS server %s...\n",
+           action == Action::add ? "Adding" : "Removing", srv);
+
+    if (action == Action::add)
+    {
+        bus.append(object.c_str(), Dbus::ethInterface, property, srv);
+    }
+    else
+    {
+        bus.remove(object.c_str(), Dbus::ethInterface, property, srv);
+    }
+
+    puts(completeMessage);
+}
+
+/** @brief Add/remove NTP server: `ntp [VLANID] {add|del} IP` */
+static void cmdNtp(Dbus& bus, Arguments& args)
+{
+    const std::string object = pathFromOptional(args);
+    const Action action = args.asAction();
+    const char* srv = args.asText();
+    args.expectEnd();
+
+    printf("%s NTP server %s...\n",
+           action == Action::add ? "Adding" : "Removing", srv);
+
+    if (action == Action::add)
+    {
+        bus.append(object.c_str(), Dbus::ethInterface, Dbus::ethNtpServers,
+                   srv);
+    }
+    else
+    {
+        bus.remove(object.c_str(), Dbus::ethInterface, Dbus::ethNtpServers,
+                   srv);
+    }
+
+    puts(completeMessage);
+}
+
+/** @brief Add/remove VLAN: `vlan {add|del} ID` */
+static void cmdVlan(Dbus& bus, Arguments& args)
+{
+    const Action action = args.asAction();
+    const uint32_t id = static_cast<uint32_t>(args.asDigit());
+    args.expectEnd();
+
+    printf("%s VLAN with ID %u...\n",
+           action == Action::add ? "Adding" : "Removing", id);
+
+    if (action == Action::add)
+    {
+        bus.call(Dbus::objectRoot, Dbus::vlanCreateInterface,
+                 Dbus::vlanCreateMethod, Dbus::eth0, id);
+    }
+    else
+    {
+        const std::string object = bus.vlanObject(id);
+        bus.call(object.c_str(), Dbus::deleteInterface, Dbus::deleteMethod);
+    }
+
+    puts(completeMessage);
+}
+
+// clang-format off
+/** @brief List of command descriptions. */
+static const Command commands[] = {
+    {"show", nullptr, "Show current configuration", cmdShow},
+    {"reset", nullptr, "Reset configuration to factory defaults", cmdReset},
+    {"mac", "MAC", "Set MAC address", cmdMac},
+    {"hostname", "NAME", "Set host name", cmdHostname},
+    {"domain", "[VLANID] {add|del} NAME", "Add or remove domain name", cmdDomain},
+    {"gateway", "IP", "Set default gateway", cmdGateway},
+    {"ip", "[VLANID] {add|del} IP[/MASK GATEWAY]", "Add or remove static IP address", cmdIp},
+    {"dhcp", "[VLANID] {enable|disable}", "Enable or disable DHCP client", cmdDhcp},
+    {"dhcpcfg", "{enable|disable} {dns|ntp}", "Enable or disable DHCP features", cmdDhcpcfg},
+    {"dns", "[VLANID] {add|del} [static] IP", "Add or remove DNS server", cmdDns},
+    {"ntp", "[VLANID] {add|del} IP", "Add or remove NTP server", cmdNtp},
+    {"vlan", "{add|del} ID", "Add or remove VLAN", cmdVlan},
```

Comment:
VLAN creation is handled poorly inside phosphor-networkd. To be specific: it removes all existing addresses on a parent interface, except for link-local, and it may result in BMC having a broken configuration if DHCP option or static configuration is not provided quickly. But it may have been changed in later version, i remember such behavior from the Vesnin versions. Have you had a chance to test VLAN creation with these changes?

Reply:
Network VLAN is not the theme I know good and I don't know how to test it. Can you help me with that?

Reply:
Sure!